### DISCUSSION TITLE
Controllers from external assemblies get added automatically in 3.1

### DISCUSSION THREAD

**Discussion by oyvindi** [dotnet/aspnetcore#25184] [General] [answered]
To my suprise, when adding a controller class in a separate assembly, it gets added automatically to the service.

For example:

1. I create a new project from scratch, that in includes the `WeatherForecastController` example
2. I create a new class library `LibWithController` in the same solution, with a `SpywareController` class
3. I add the `LibWithController` as a project reference to the application
4. The service now automagically exposes the `WeatherForecastController` and `SpywareController` functions (also shows in Swagger)

In other words, the controller mapper seems to traverse every dependency and adds whatever controller that comes in its way.

In 3.0 and earlier, I think one had to something like this:

`services.AddControllers().PartManager.ApplicationParts.Add(new AssemblyPart(assembly));`

Is this intentional behaviour ?

Attached example here:
[AspCoreSandbox.zip](https://github.com/dotnet/aspnetcore/files/5116314/AspCoreSandbox.zip)

**Comment by javiercn** [comment, marked as answer]
@oyvindi thanks for contacting us.

This has been the behavior since asp.net core 1.0. I'm not sure why you had to add the part manually in your case, other than you explicitly disabled the dependencycontext generation.
